### PR TITLE
Fix InfiniteScroll loop when Data length changes

### PR DESCRIPTION
**Pull request by leossantiago**
#### What does this PR do?

Introduces a patch to #5095,  introduced on PR #4936 .

This fix uses two states to help tracking changes to data length. On this approach, we avoid going back to using `items.length === 0` as a check and thus we wouldn't assume that the caller wants to reset the data to an empty array before repopulating it.

Please note that this covers a very specific use case ( #4934 ) where the reporter wanted to reset data to an empty array before setting it. The condition `lastPage === 0` will make this work only for situations where the modified length is smaller than the initial step. 

Would love to hear your thoughts on that @ShimiSun @halocline, and I'd suggest we open this matter to discussion so we can have a clear view of use cases we want to cover before any attempt of broadening the whole reset-onMore-calls thing.

#### Where should the reviewer start?

`src/js/components/InfiniteScroll/InfiniteScroll.js`

#### What testing has been done on this PR?

Manual testing as JSDom and Jest do not support calls such as getBoundingClientRect, making this PR unable to be tested using automated tests.

#### How should this be manually tested?

Code provided on PR #4936 and issue #5095.

#### What are the relevant issues?

#5095

#### Do the grommet docs need to be updated?

No.

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible.

### CODE REVIEW

**Comment by halocline** [comment]
Thank you @leossantiago. Much appreciated. I will check this out within the next hour. 

I agree that discussion should happen around the use cases we want to support.

**Comment by halocline** [comment]
> Would love to hear your thoughts on that @ShimiSun @halocline, and I'd suggest we open this matter to discussion so we can have a clear view of use cases we want to cover before any attempt of broadening the whole reset-onMore-calls thing.
> 
💯  We definitely need to discuss the scenarios we want to support.

> On this approach, we avoid going back to using `items.length === 0` as a check and thus we wouldn't assume that the caller wants to reset the data to an empty array before repopulating it.
> 
> Please note that this covers a very specific use case ( #4934 ) where the reporter wanted to reset data to an empty array before setting it. The condition `lastPage === 0` will make this work only for situations where the modified length is smaller than the initial step.
> 

@leossantiago This approach works around the assumption of wanting to reset to an empty array, but to your point about:
> this [will only] work for situations where the modified length is smaller than the initial step.
> 

In your opinion, with the constraint of `items.length < step`, does the incremental value of not assuming an empty array justify the cost of the additional complexity? Would we be better off keeping things simple for now using the `items.length === 0` approach and then having a proper use case and design discussion as follow up?

**Comment by leossantiago** [comment]
I believe using `items.length === 0` would be the best option for now. The whole point of having the `useEffect` introduced on #4936 was to deal with `onMore` not working as intended when resetting the data passed to `DataTable`. Even though we are trying to find a solution that works for any changes on data, we are still bound to having a `lastPage` equal to zero in order for this to work.

> and then having a proper use case and design discussion

Definitely. This was introduced based on a very specific case. We need to figure out what we want to have as general behavior.. 
I wonder if removing the `lastPage` condition would do the trick, but I'm worried this could lead us to another loop.

**Comment by halocline** [comment]
Appreciate you sharing your thoughts.

> I believe using `items.length === 0` would be the best option for now. 

I agree and would support moving forward with this direction to get the fix in. We can then discuss how/whether we want to enhance the onMore/reset concept.

Would you be willing to modify this PR to head in this direction? Or, I am have ability this afternoon and would be happy to open a separate PR.

**Comment by leossantiago** [comment]
> Would you be willing to modify this PR to head in this direction? 

Sure thing. Will do that in a couple of minutes.

**Comment by leossantiago** [comment]
@halocline done! Just force pushed an update to this PR.

File: src/js/components/InfiniteScroll/InfiniteScroll.js [hunk 1]
```diff
@@ -151,12 +151,12 @@ const InfiniteScroll = ({
   }, [items.length, lastPage, onMore, pendingLength, renderPageBounds, step]);
 
   useEffect(() => {
-    if (lastPage === 0 && pendingLength !== 0) {
+    if (items.length && lastPage === 0 && pendingLength !== 0) {
```
Comment (halocline) [inline src/js/components/InfiniteScroll/InfiniteScroll.js:154]: Think we lost the `=== 0`

```suggestion
    if (items.length === 0 && lastPage === 0 && pendingLength !== 0) {
```

Comment (leossantiago) [reply src/js/components/InfiniteScroll/InfiniteScroll.js:154]: Oops. My bad. Will fix it in a sec!

**Review by halocline — APPROVED** [review]
Thank you. Again, appreciate you working through this with us.

Tested manually with the following stories:
[Select + onMore](https://gist.github.com/halocline/d8d7d11aca684eaa624f840d6c703106)

[InfiniteScroll + onMore + reset](https://gist.github.com/halocline/f805d07559981dda24a19f870d6666bf)

**Comment by leossantiago** [comment]
Np, @halocline! Feels great to be part of this community! Glad to be helpful 😄

**Review by ShimiSun — APPROVED** [review]
Thank you @leossantiago ! Enviando amor do Colorado!

**Pull request merged**